### ISSUE TITLE
Set Lifestyle of the resulting collection using RegisterCollection?

### ISSUE THREAD

**Issue by apetkovic** [simpleinjector/SimpleInjector#152]
I register most of my dependencies using the `WebRequestLifestyle` in an MVC application. Some of those dependencies require a collection to be injected into them and I register those collections with `RegisterCollection`. However, when I run my application I get an exception that

> A lifestyle mismatch is encountered. UserManager (Web Request) depends on `IPushProvider[]` (Transient).

It appears the collection itself is transient. I've set my container options of `DefaultScopedLifestyle = WebRequestLifestyle` and `LifestyleSelectionBehavior = ScopedLifestyleSelectionBehavior` but that didn't help. How can I set the lifestyle of the collection itself, not just types within the collection?

**Comment by dotnetjunkie** [comment]
In the case of working with arrays, Simple Injector will always create a new (transient) array for you, even though all the elements of the array have a larger lifestyle. This will cause Simple Injector's verification to fail in case a non-transient component depends on an array.

There are two reasons why Simple Injector registers arrays as transient (even if you override the `LifestyleSelectionBehavior`):
 - Since users can define custom lifestyles, it can become impossible to select right lifestyle for the array. This can accidentally cause elements to live longer than intended.
 - Arrays are mutable and can be changed by anyone who uses it. Injecting an array with a lifestyle longer than transient could cause other components to see the accidental changes to the array and might cause the application to break in a seemingly unrelated place.

The solution to your problem is actually very simple. Instead of injecting an array, inject one of the following types:

 - `IList<T>`
 - `ICollection<T>`
 - `IReadOnlyList<T>`
 - `IReadOnlyCollection<T>`
 - `IEnumerable<T>`

In case of any of the given types, Simple Injector will not inject a Transient object but a *Singleton*. This will not cause any problems even if the services in the collections are not singleton, because the injected object behaves as a stream. This means that every time you iterate it, the stream will call back into the container to get the item by its correct lifestyle.

**Comment by apetkovic** [comment]
Wow, thanks for the super fast reply! I've changed my business logic layer to use IEnumerable<T> in the constructors instead of arrays and that certainly worked. Thanks! This is an unrelated question, and perhaps I should post it in a new thread... I've used OnActivation and OnDeactivation methods in Ninject to execute logic when an instance is first resolved. I checked the documentation and couldn't find anything like that for Simple Injector. Is there a way to extend Simple Injector and create this behavior?
